File: Ex02-Bottlenecks/thread_bottlenecks.py
import threading 

i = 0

lock = threading.Lock()

def thread_func1():
	lock.acquire()
	global i
	for j in range(0, 1000000):
		i += 1
	lock.release()

def thread_func2():
	lock.acquire()
	global i
	for j in range(1, 1000000):
		i -= 1
	lock.release()

def main():
	print("Before thread join: " + str(i))

	thread_1 = threading.Thread(target = thread_func1, args = (),)
	thread_2 = threading.Thread(target = thread_func2, args = (),)
	thread_1.start()
	thread_2.start()

	thread_1.join()
	thread_2.join()
	
	print("After thread join: " + str(i))

main()

File: Ex02-Bottlenecks/C_code/thread_bottlenecks.c
//haakoneh & erlendvd

#include <pthread.h>
#include <stdio.h>


int i = 0;

pthread_mutex_t lock;

void* thread_func_1(){
	

	unsigned int j;
	for(j = 0; j < 1000000; j++){
		pthread_mutex_lock(&lock);
		i++;
		pthread_mutex_unlock(&lock);
	}

	

	return NULL;
}

void* thread_func_2(){
	unsigned int j;
	for(j = 1; j < 1000000; j++){
		pthread_mutex_lock(&lock);
		i--;
		pthread_mutex_unlock(&lock);
	}


	return NULL;
}

int main(){
    pthread_t thread_1;
    pthread_t thread_2;

    pthread_create(&thread_1, NULL, thread_func_1, NULL);
    pthread_create(&thread_2, NULL, thread_func_2, NULL);

    printf("Before joined threads:  %i\n", i);

    pthread_join(thread_1, NULL);
    pthread_join(thread_2, NULL);

    printf("After joined threads:  %i\n", i);

    return 0;
}
